pkg/operator/controller/crl/controller.go

```diff
@@ -286,12 +286,14 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 
 	// TODO Consider letting ensureCRLConfigmap get the deployment and build
 	// the owner reference as we don't know yet whether we need it.
-	if _, _, err := r.ensureCRLConfigmap(ctx, ic, deployment.Namespace, ownerRef, haveCAConfigmap, clientCAConfigmap); err != nil {
+	if _, _, ctx, err := r.ensureCRLConfigmap(ctx, ic, deployment.Namespace, ownerRef, haveCAConfigmap, clientCAConfigmap); err != nil {
 		return reconcile.Result{}, fmt.Errorf("failed to ensure client CA CRL configmap for ingresscontroller %s: %w", request.NamespacedName, err)
+	} else {
+		if nextCRLUpdate, ok := ctx.Value("nextCRLUpdate").(time.Time); ok && !nextCRLUpdate.IsZero() {
+			log.Info("Requeueing when next CRL expires", "requeue time", nextCRLUpdate.String(), "time until requeue", time.Until(nextCRLUpdate))
+			//Re-reconcile when any of the CRLs expire
+			return reconcile.Result{RequeueAfter: time.Until(nextCRLUpdate)}, nil
```

Comment:
Thoughts?

Reply:
Thoughts?

Reply:
Sorry it took me a while to get back to you on this one. I think that's a valid concern. I'm not sure at first glance what's the best way to solve it, but let me look into this.

Reply:
@gcs278 I've pushed a change that seems to fix this from my manual testing; please take a look when you get a chance and let me know what you think.

I added a global variable to track when the next CRL update is expected to be, and now the reconcile is only triggered when the next update as computed in `desiredCRLConfigMap` will be sooner than any already pending requeues.

Reply:
> Is there a risk of this requeueing multiple times for the same next CRL update? I.e. on every reconcile, do we queue up this request? Is that okay?

Are you worried that the queue may never be empty?  I think it is all right.  Is there some other potential problem with the logic as it was when you asked about it?

Reply:
> Are you worried that the queue may never be empty?

I don't think that's what I meant. I just noticed we queue up a request on every reconcile for the same next CRL update. That's not a big problem, just seemed a bit extraneous. We might have 10 or 20 requests that will all trigger at the same time for the same (or immediately after each other) nextCRLUpdate, depending on how many times the CRL was reconciled.

I don't know if there is an easy solution to say "this request is already in the queue, don't queue again". Just pointing it out more than anything.

Reply:
I was able to do a bit more testing on this, and I got the operator to requeue a reconcile for the same time 10 times, but when the time came to do the reconcile, it seems to execute the reconcile twice then stop retrying (at least for a bit; when the new CRL expired 5 minutes later it still re-reconciled). I'm not sure why it's twice; maybe there's a rate limit implemented somewhere, but it doesn't look like this will flood the operator

Reply:
Yeah, I think we're all right here.  I did a little more investigation: 

* controller-runtime's `reconcileHandler` calls our `Reconcile` method (technically, it calls the controller's `Reconcile` method, but that ultimately results in ours' being called) and requeues the request using `AddAfter` if the result has nil error and non-zero `RequeueAfter`: https://github.com/openshift/cluster-ingress-operator/blob/362a3ec2a92f501477dff3ec71348c8150437709/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go#L320-L333
* `AddAfter` adds the item to the queue using `Add` or sends a `waitFor` to a channel: https://github.com/openshift/cluster-ingress-operator/blob/362a3ec2a92f501477dff3ec71348c8150437709/vendor/k8s.io/client-go/util/workqueue/delaying_queue.go#L161-L179
* The queue has a goroutine that reads the `waitFor` off the channel and adds it to its internal `waitingForQueue` queue using `insert` or adds it to the queue using `Add`: https://github.com/openshift/cluster-ingress-operator/blob/362a3ec2a92f501477dff3ec71348c8150437709/vendor/k8s.io/client-go/util/workqueue/delaying_queue.go#L253-L257
* `insert` checks whether there is an existing entry for the reconcile request; if there is one, it updates the existing entry in the queue instead of pushing a duplicate entry onto the queue: https://github.com/openshift/cluster-ingress-operator/blob/362a3ec2a92f501477dff3ec71348c8150437709/vendor/k8s.io/client-go/util/workqueue/delaying_queue.go#L267-L277
* The queue itself defines `Add` to prevent adding duplicates: https://github.com/openshift/cluster-ingress-operator/blob/362a3ec2a92f501477dff3ec71348c8150437709/vendor/k8s.io/client-go/util/workqueue/queue.go#L119-L137

Between Ryan's testing and my understanding of the queue and delaying_queue logic, I believe that there should be no issue with duplicate items in the queue or the delaying_queue's internal channel or queue.